src/aux-records/XpController.spec.ts

```diff
@@ -0,0 +1,147 @@
+import { createTestControllers, createTestUser } from './TestUtils';
+import { MemoryAuthMessenger } from './MemoryAuthMessenger';
+import { AuthController } from './AuthController';
+import { MemoryStore } from './MemoryStore';
+import { GetXpUserResult, XpController } from '../aux-records/XpController';
+import { XpUser } from './XpStore';
+import { isOfXType } from './Utils';
+import { PromiseT } from './TypeUtils';
+
+/**
+ * XpController tests
+ */
+describe('XpController', () => {
+    //* BOCS: Init services
+    let services: ReturnType<typeof createTestControllers>;
+    let xpController: XpController;
+    let memoryStore: MemoryStore;
+    let authController: AuthController;
+    let authMessenger: MemoryAuthMessenger;
+    //* EOCS: Init services
+
+    /**
+     * (Re-)Initialize services required for tests
+     */
+    const initServices = () => {
+        services = createTestControllers();
+        xpController = new XpController({
+            xpStore: services.store,
+            authController: services.auth,
+            authStore: services.store,
+        });
+        memoryStore = services.store;
+        authController = services.auth;
+        authMessenger = services.authMessenger;
+    };
+
+    /**
+     * Runs before all tests (once)
+     */
+    beforeAll(async () => {
+        initServices();
+    });
+
+    /**
+     * Runs before each test
+     */
+    beforeEach(() => {
+        initServices();
+    });
+
+    //* Test case for the XpController class
+    it('should be defined', () => {
+        expect(XpController).toBeTruthy();
+        expect(xpController).toBeTruthy();
+    });
+
+    /**
+     * Successful getXpUser invocations across the board should be indicative of—
+     * successful XP user initialization as well as retrieval
+     */
+    describe('getXpUser', () => {
+        //* An auth user for use in testing getXpUser
+        let _authUser: PromiseT<ReturnType<typeof createTestUser>>;
+
+        //* The expected result of a successful getXpUser call
+        const successXpUser: GetXpUserResult = {
+            success: true,
+            user: {
+                id: expect.any(String),
+                userId: expect.any(String),
+                accountId: expect.any(String),
+                requestedRate: null,
+                createdAtMs: expect.any(Number),
+                updatedAtMs: expect.any(Number),
```

Comment:
@LandonSiler Generally, it is best to prefer exact values (if they're able to be known) over expecting "any" value. This will help the tests be more specific, and also help us catch scenarios where we accidentally break the tests or change some behavior. This means more "duplicate" code in tests, but tests are here in part to make sure that things don't change in wrong ways, and this means that copying code within tests is usually ok.

As an extra note, if you have a scenario where you need to control the output of `uuid()` or `Date.now()`, you can use mock functions. AuthController.spec.ts has an example of how to setup mocks for `uuid()` in [line 75](https://github.com/casual-simulation/casualos/blob/develop/src/aux-records/AuthController.spec.ts#L75), [line 225](https://github.com/casual-simulation/casualos/blob/develop/src/aux-records/AuthController.spec.ts#L225), and [line 260](https://github.com/casual-simulation/casualos/blob/develop/src/aux-records/AuthController.spec.ts#L260)

Reply:
I have literal matching in non-commited contract testing ha!
I will correct this, the main reason I believe it slipped past me refactoring it initially was because it's declared outside of a specific test scope which made matching literal values require parameterization when using something like a template.